File: Java/_257_Binary_Tree_Paths/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> result = new ArrayList<String>();
        if (root == null) return result;
        dfs(root, result, "");
        return result;
    }

    private void dfs(TreeNode root, List<String> result, String path) {
        if (root == null) return;
        path += (path.length() == 0 ? "" : "->") + root.val;
        if (root.left == null && root.right == null) {
            result.add(path);
            return;
        }
        dfs(root.left, result, path);
        dfs(root.right, result, path);
    }
}

File: Python/_257_Binary_Tree_Paths/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def binaryTreePaths(self, root):
        if root == None:
            return []
        return self.dfs(root, str(root.val), [])

    def dfs(self, root, val, paths):
        if not root.left and not root.right:
            paths.append(val)
        if root.left:
            paths = self.dfs(root.left, val + "->" + str(root.left.val), paths)
        if root.right:
            paths = self.dfs(root.right, val + "->" + str(root.right.val), paths)
        return paths

